tensorflow/compiler/jit/deadness_analysis.cc

```diff
@@ -914,10 +954,135 @@ Status GetFullFrame(const Node* n, absl::Span<const ControlFlowInfo> cfi_infos,
 
   return Status::OK();
 }
+
+// If the node is inside some frames, get the name of the outermost non-empty
+// frame.  Otherwise, get an empty frame name.
+Status GetRootFrame(const Node* n, absl::Span<const ControlFlowInfo> cfi_infos,
+                    absl::string_view* frame) {
+  int depth = 0;
+  const ControlFlowInfo* cfi_iter = &cfi_infos[n->id()];
+  while (!cfi_iter->parent_frame->IsSource()) {
+    n = cfi_iter->parent_frame;
+    cfi_iter = &cfi_infos[n->id()];
+
+    if (depth++ > 5000) {
+      return errors::Internal(
+          "Frame of depth > 5000:  Probably malformed graph or a bug in "
+          "BuildControlFlowInfo");
+    }
+  }
+
+  *frame = cfi_iter->frame_name;
+  return Status::OK();
+}
+
+// Compute a special topological order for the Graph, where nodes having the
+// same root frame are placed adjacent to each other.  The traversal uses a
+// variant of Kahn's algorithm.  num_ready_inputs is used to keep track of how
+// many inputs of each node are ready; a node is ready to be scheduled if all
+// of its inputs are ready.
+// Ref. to https://en.wikipedia.org/wiki/Topological_sorting for details.
+Status GetFrameBasedTopologicalOrder(const Graph* g,
+                                     absl::Span<const ControlFlowInfo> cf_infos,
+                                     std::vector<Node*>* order) {
+  absl::flat_hash_map<absl::string_view, size_t> num_enters;
+  absl::flat_hash_map<absl::string_view, size_t> num_exits;
+  std::vector<size_t> num_ready_inputs(g->num_node_ids(), 0);
+  Node* src_node = g->source_node();
+  for (const auto* node : g->op_nodes()) {
```

Comment:
Might be better to extract out the first half of this loop into a `GetEntryAndExitsPerRootFrame` helper.

Reply:
I don't see a clear way to do that. Currently, the loop counts 3 things (`num_enters`, `num_exits`, and `num_ready_inputs`.) in one traversal. Not easy to factor out part of it.

Reply:
I was thinking we could do two traversals (more expensive but the code is clearer), but this is fine for now.